src/networks/invertible_network_conditional_glow.jl

```diff
@@ -0,0 +1,181 @@
+
+ - `K`: number of flow steps per scale (inner loop)
+
+ - `split_scales`: if true, perform squeeze operation which halves spatial dimensions and duplicates channel dimensions
+    then split output in half along channel dimension after each scale. Feed one half through the next layers,
+    while saving the remaining channels for the output.
+
+ - `k1`, `k2`: kernel size of convolutions in residual block. `k1` is the kernel of the first and third 
+ operator, `k2` is the kernel size of the second operator.
+
+ - `p1`, `p2`: padding for the first and third convolution (`p1`) and the second convolution (`p2`)
+
+ - `s1`, `s2`: stride for the first and third convolution (`s1`) and the second convolution (`s2`)
+
+ - `ndims` : number of dimensions
+
+ - `squeeze_type` : squeeze type that happens at each multiscale level
+
+ *Output*:
+ 
+ - `G`: invertible Glow network.
+
+ *Usage:*
+
+ - Forward mode: `Y, logdet = G.forward(X)`
+
+ - Backward mode: `ΔX, X = G.backward(ΔY, Y)`
+
+ *Trainable parameters:*
+
+ - None in `G` itself
+
+ - Trainable parameters in activation normalizations `G.AN[i,j]` and coupling layers `G.C[i,j]`,
+   where `i` and `j` range from `1` to `L` and `K` respectively.
+
+ See also: [`ActNorm`](@ref), [`CouplingLayerGlow!`](@ref), [`get_params`](@ref), [`clear_grad!`](@ref)
+"""
+struct NetworkConditionalGlow <: InvertibleNetwork
+    AN::AbstractArray{ActNorm, 2}
+    AN_C::ActNorm
+    CL::AbstractArray{ConditionalLayerGlow, 2}
+    Z_dims::Union{Array{Array, 1}, Nothing}
+    L::Int64
+    K::Int64
+    squeezer::Squeezer
+    split_scales::Bool
+end
+
+@Flux.functor NetworkConditionalGlow
+
+# Constructor
+function NetworkConditionalGlow(n_in, n_cond, n_hidden, L, K; split_scales=false, rb_activation::ActivationFunction=ReLUlayer(), k1=3, k2=1, p1=1, p2=0, s1=1, s2=1, ndims=2, squeezer::Squeezer=ShuffleLayer(), activation::ActivationFunction=SigmoidLayer())
+    AN = Array{ActNorm}(undef, L, K)    # activation normalization
+    AN_C = ActNorm(n_cond; logdet=false)    # activation normalization for condition
+    CL = Array{ConditionalLayerGlow}(undef, L, K)  # coupling layers w/ 1x1 convolution and residual block
+ 
+    if split_scales
+        Z_dims = fill!(Array{Array}(undef, L-1), [1,1]) #fill in with dummy values so that |> gpu accepts it   # save dimensions for inverse/backward pass
+        channel_factor = 4
+    else
+        Z_dims = nothing
+        channel_factor = 1
+    end
+
+    for i=1:L
+        n_in *= channel_factor # squeeze if split_scales is turned on
+        n_cond *= channel_factor # squeeze if split_scales is turned on
+        for j=1:K
+            AN[i, j] = ActNorm(n_in; logdet=true)
+            CL[i, j] = ConditionalLayerGlow(n_in, n_cond, n_hidden; rb_activation=rb_activation, k1=k1, k2=k2, p1=p1, p2=p2, s1=s1, s2=s2, logdet=true, activation=activation, ndims=ndims)
+        end
+        (i < L && split_scales) && (n_in = Int64(n_in/2)) # split
+    end
```

Comment:
Like I know this is how we've setup all these different one but isn't this just `Chain( Chain(squeezer, AN, Glow) ... )` ?

That'd cleanup so much code duplicate to only define the base layers and then build network as a chain of layer like in Flux/torch/..... . Would be much more flexible as well too.

but I get it's a lot of work so maybe for later

Reply:
That would definitely work for some of the networks. But I have a hard time seeing how to use more involved networks like for the split_scales option where you need to split the latent vector in half, save one half, save the dims to undo this, etc etc.